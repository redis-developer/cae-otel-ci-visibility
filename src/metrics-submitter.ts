import {
  metrics,
  type MetricOptions,
  type Histogram,
  type Counter,
  type UpDownCounter
} from '@opentelemetry/api'
import type { MeterProvider } from '@opentelemetry/sdk-metrics'
import type { TMetricDataPoint, TMetricsConfig } from './metrics-generator.js'

type TMetric = Histogram | Counter | UpDownCounter

export class MetricsSubmitter {
  private readonly meter
  private readonly histograms = new Map<string, Histogram>()
  private readonly counters = new Map<string, Counter>()
  private readonly upDownCounters = new Map<string, UpDownCounter>()
  private readonly namespace
  private readonly version

  constructor(
    config: TMetricsConfig,
    meterProvider: MeterProvider | undefined,
    namespace: string,
    version: string
  ) {
    if (meterProvider) {
      metrics.disable()
      metrics.setGlobalMeterProvider(meterProvider)
    }

    this.namespace = namespace
    this.version = version
    this.meter = metrics.getMeter(config.serviceName, config.serviceVersion)
  }

  public submitMetrics(metricDataPoints: readonly TMetricDataPoint[]): void {
    for (const dataPoint of metricDataPoints) {
      switch (dataPoint.metricType) {
        case 'histogram':
          this.recordHistogram(dataPoint)
          break
        case 'counter':
          this.incrementCounter(dataPoint)
          break
        case 'updowncounter':
          this.updateUpDownCounter(dataPoint)
          break
      }
    }
  }

  private getOrCreateMetric<T extends TMetric>(
    metricName: string,
    metricMap: Map<string, T>,
    createMetric: () => T
  ): T {
    if (!metricMap.has(metricName)) {
      metricMap.set(metricName, createMetric())
    }

    const metric = metricMap.get(metricName)
    if (!metric) {
      throw new Error(`Not found: ${metricName}`)
    }
    return metric
  }

  private createHistogramOptions(dataPoint: TMetricDataPoint): MetricOptions {
    const options: MetricOptions = {
      description: dataPoint.description,
      unit: dataPoint.unit
    }

    if (dataPoint.buckets) {
      options.advice = {
        explicitBucketBoundaries: dataPoint.buckets.concat()
      }
    }

    return options
  }

  private recordHistogram(dataPoint: TMetricDataPoint): void {
    const histogram = this.getOrCreateMetric(
      dataPoint.metricName,
      this.histograms,
      () =>
        this.meter.createHistogram(
          `${this.namespace}.${this.version}.${dataPoint.metricName}`,
          this.createHistogramOptions(dataPoint)
        )
    )

    histogram.record(dataPoint.value, dataPoint.attributes)
  }

  private incrementCounter(dataPoint: TMetricDataPoint): void {
    const counter = this.getOrCreateMetric(
      dataPoint.metricName,
      this.counters,
      () =>
        this.meter.createCounter(
          `${this.namespace}.${this.version}.${dataPoint.metricName}`,
          {
            description: dataPoint.description,
            unit: dataPoint.unit
          }
        )
    )

    counter.add(dataPoint.value, dataPoint.attributes)
  }

  private updateUpDownCounter(dataPoint: TMetricDataPoint): void {
    const upDownCounter = this.getOrCreateMetric(
      dataPoint.metricName,
      this.upDownCounters,
      () =>
        this.meter.createUpDownCounter(
          `${this.namespace}.${this.version}.${dataPoint.metricName}`,
          {
            description: dataPoint.description,
            unit: dataPoint.unit
          }
        )
    )

    upDownCounter.add(dataPoint.value, dataPoint.attributes)
  }
}
