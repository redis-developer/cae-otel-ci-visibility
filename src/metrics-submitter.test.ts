import { MetricsSubmitter } from './metrics-submitter.js'
import type { TMetricDataPoint, TMetricsConfig } from './metrics-generator.js'
import { MeterProvider } from '@opentelemetry/sdk-metrics'
import {
  InMemoryMetricExporter,
  ResourceMetrics
} from '@opentelemetry/sdk-metrics'

import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics'
import { AggregationTemporality } from '@opentelemetry/sdk-metrics'
import { resourceFromAttributes } from '@opentelemetry/resources'
import { setTimeout } from 'timers/promises'

const testResource = resourceFromAttributes({
  'service.name': 'test-service',
  'service.version': '1.0.0'
})

async function waitForNumberOfExports(
  exporter: InMemoryMetricExporter,
  numberOfExports: number
): Promise<ResourceMetrics[]> {
  if (numberOfExports <= 0) {
    throw new Error('numberOfExports must be greater than or equal to 0')
  }

  let totalExports = 0
  let attempts = 0
  const maxAttempts = 50

  while (totalExports < numberOfExports && attempts < maxAttempts) {
    await setTimeout(20)
    const exportedMetrics = exporter.getMetrics()
    totalExports = exportedMetrics.length
    attempts++
  }

  if (attempts >= maxAttempts) {
    throw new Error(
      `Timeout waiting for ${numberOfExports} exports after ${maxAttempts} attempts`
    )
  }
  return exporter.getMetrics()
}

function normalizeMetricsForSnapshot(
  exportedMetrics: ResourceMetrics[]
): unknown[] {
  return exportedMetrics.map((resourceMetric) => ({
    ...resourceMetric,
    scopeMetrics: resourceMetric.scopeMetrics.map((scopeMetric) => ({
      ...scopeMetric,
      metrics: scopeMetric.metrics.map((metric) => ({
        ...metric,
        dataPoints: metric.dataPoints.map((dataPoint) => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { endTime, startTime, ...rest } = dataPoint

          return rest
        })
      }))
    }))
  }))
}

type TTestCase = {
  name: string
  metrics: readonly TMetricDataPoint[]
  expectExports?: number
}

describe('MetricsSubmitter', () => {
  let exporter: InMemoryMetricExporter
  let meterProvider: MeterProvider
  let metricReader: PeriodicExportingMetricReader
  let submitter: MetricsSubmitter

  const config: TMetricsConfig = {
    serviceName: 'test-service',
    serviceVersion: '1.0.0',
    environment: undefined,
    repository: undefined,
    branch: undefined,
    commitSha: undefined,
    runId: undefined,
    jobUUID: undefined
  }

  const createDataPoint = (
    overrides: Partial<TMetricDataPoint> = {}
  ): TMetricDataPoint => ({
    metricName: 'test.metric',
    metricType: 'counter',
    value: 1,
    attributes: { 'test.name': 'example' },
    description: 'Test metric',
    unit: '{test}',
    ...overrides
  })

  beforeEach(() => {
    exporter = new InMemoryMetricExporter(AggregationTemporality.CUMULATIVE)

    metricReader = new PeriodicExportingMetricReader({
      exporter: exporter,
      exportIntervalMillis: 100,
      exportTimeoutMillis: 50
    })

    meterProvider = new MeterProvider({
      resource: testResource,
      readers: [metricReader]
    })
    submitter = new MetricsSubmitter(
      config,
      meterProvider,
      'test-namespace',
      'v1'
    )
  })

  afterEach(async () => {
    await exporter.shutdown()
    await metricReader.shutdown()
  })

  const testCases: readonly TTestCase[] = [
    {
      name: 'creates instruments for each metric type',
      metrics: [
        createDataPoint({
          metricName: 'test.duration',
          metricType: 'histogram',
          unit: 's'
        }),
        createDataPoint({
          metricName: 'test.status',
          metricType: 'counter',
          unit: '{test}'
        }),
        createDataPoint({
          metricName: 'test.suite.total',
          metricType: 'updowncounter',
          unit: '{test}'
        })
      ]
    },
    {
      name: 'configures histogram with explicit buckets',
      metrics: [
        createDataPoint({
          metricName: 'test.duration',
          metricType: 'histogram'
        })
      ]
    },
    {
      name: 'reuses existing instruments for same metric name',
      metrics: [
        createDataPoint({ metricName: 'test.counter', value: 1 }),
        createDataPoint({ metricName: 'test.counter', value: 2 })
      ]
    },
    {
      name: 'handles multiple metric types in one submission',
      metrics: [
        createDataPoint({
          metricName: 'test.histogram',
          metricType: 'histogram',
          value: 42
        }),
        createDataPoint({
          metricName: 'test.counter',
          metricType: 'counter',
          value: 10
        }),
        createDataPoint({
          metricName: 'test.updown',
          metricType: 'updowncounter',
          value: -5
        })
      ]
    },
    {
      name: 'handles metrics with different attributes correctly',
      metrics: [
        createDataPoint({
          metricName: 'test.counter',
          attributes: { env: 'dev', team: 'alpha' },
          value: 5
        }),
        createDataPoint({
          metricName: 'test.counter',
          attributes: { env: 'prod', team: 'beta' },
          value: 3
        })
      ]
    },
    {
      name: 'handles updowncounter negative values correctly',
      metrics: [
        createDataPoint({
          metricName: 'test.updown',
          metricType: 'updowncounter',
          value: 10
        }),
        createDataPoint({
          metricName: 'test.updown',
          metricType: 'updowncounter',
          value: -15
        })
      ]
    }
  ]

  for (const testCase of testCases) {
    it(`should ${testCase.name}`, async () => {
      submitter.submitMetrics(testCase.metrics)
      await metricReader.forceFlush()
      const expectedExports = testCase.expectExports ?? 1
      const exportedMetrics = await waitForNumberOfExports(
        exporter,
        expectedExports
      )
      expect(exportedMetrics.length).toBe(expectedExports)

      const normalizedMetrics = normalizeMetricsForSnapshot(exportedMetrics)
      expect(normalizedMetrics).toMatchSnapshot()
    })
  }

  it('should handle empty metrics array gracefully', async () => {
    submitter.submitMetrics([])
    expect(submitter).toBeDefined()
    await metricReader.forceFlush()
    const exportedMetrics = exporter.getMetrics()
    expect(exportedMetrics).toHaveLength(0)
  })
})
