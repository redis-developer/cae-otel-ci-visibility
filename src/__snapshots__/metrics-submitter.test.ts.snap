// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MetricsSubmitter should configures histogram with explicit buckets 1`] = `
[
  {
    "resource": ResourceImpl {
      "_asyncAttributesPending": false,
      "_memoizedAttributes": undefined,
      "_rawAttributes": [
        [
          "service.name",
          "test-service",
        ],
        [
          "service.version",
          "1.0.0",
        ],
      ],
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 0,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": {
                  "buckets": {
                    "boundaries": [
                      0.1,
                      0.5,
                      1,
                      5,
                      10,
                    ],
                    "counts": [
                      0,
                      0,
                      1,
                      0,
                      0,
                      0,
                    ],
                  },
                  "count": 1,
                  "max": 1,
                  "min": 1,
                  "sum": 1,
                },
              },
            ],
            "descriptor": {
              "advice": {
                "explicitBucketBoundaries": [
                  0.1,
                  0.5,
                  1,
                  5,
                  10,
                ],
              },
              "description": "Test metric",
              "name": "test-namespace.v1.test.duration",
              "type": "HISTOGRAM",
              "unit": "{test}",
              "valueType": 1,
            },
          },
        ],
        "scope": {
          "name": "test-service",
          "schemaUrl": undefined,
          "version": "1.0.0",
        },
      },
    ],
  },
]
`;

exports[`MetricsSubmitter should creates instruments for each metric type 1`] = `
[
  {
    "resource": ResourceImpl {
      "_asyncAttributesPending": false,
      "_memoizedAttributes": undefined,
      "_rawAttributes": [
        [
          "service.name",
          "test-service",
        ],
        [
          "service.version",
          "1.0.0",
        ],
      ],
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 0,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": {
                  "buckets": {
                    "boundaries": [
                      0,
                      5,
                      10,
                      25,
                      50,
                      75,
                      100,
                      250,
                      500,
                      750,
                      1000,
                      2500,
                      5000,
                      7500,
                      10000,
                    ],
                    "counts": [
                      0,
                      1,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                    ],
                  },
                  "count": 1,
                  "max": 1,
                  "min": 1,
                  "sum": 1,
                },
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.duration",
              "type": "HISTOGRAM",
              "unit": "s",
              "valueType": 1,
            },
          },
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": 1,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.status",
              "type": "COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": true,
          },
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": 1,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.suite.total",
              "type": "UP_DOWN_COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": false,
          },
        ],
        "scope": {
          "name": "test-service",
          "schemaUrl": undefined,
          "version": "1.0.0",
        },
      },
    ],
  },
]
`;

exports[`MetricsSubmitter should handles metrics with different attributes correctly 1`] = `
[
  {
    "resource": ResourceImpl {
      "_asyncAttributesPending": false,
      "_memoizedAttributes": undefined,
      "_rawAttributes": [
        [
          "service.name",
          "test-service",
        ],
        [
          "service.version",
          "1.0.0",
        ],
      ],
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "env": "dev",
                  "team": "alpha",
                },
                "value": 5,
              },
              {
                "attributes": {
                  "env": "prod",
                  "team": "beta",
                },
                "value": 3,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.counter",
              "type": "COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "test-service",
          "schemaUrl": undefined,
          "version": "1.0.0",
        },
      },
    ],
  },
]
`;

exports[`MetricsSubmitter should handles multiple metric types in one submission 1`] = `
[
  {
    "resource": ResourceImpl {
      "_asyncAttributesPending": false,
      "_memoizedAttributes": undefined,
      "_rawAttributes": [
        [
          "service.name",
          "test-service",
        ],
        [
          "service.version",
          "1.0.0",
        ],
      ],
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 0,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": {
                  "buckets": {
                    "boundaries": [
                      0,
                      5,
                      10,
                      25,
                      50,
                      75,
                      100,
                      250,
                      500,
                      750,
                      1000,
                      2500,
                      5000,
                      7500,
                      10000,
                    ],
                    "counts": [
                      0,
                      0,
                      0,
                      0,
                      1,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                      0,
                    ],
                  },
                  "count": 1,
                  "max": 42,
                  "min": 42,
                  "sum": 42,
                },
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.histogram",
              "type": "HISTOGRAM",
              "unit": "{test}",
              "valueType": 1,
            },
          },
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": 10,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.counter",
              "type": "COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": true,
          },
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": -5,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.updown",
              "type": "UP_DOWN_COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": false,
          },
        ],
        "scope": {
          "name": "test-service",
          "schemaUrl": undefined,
          "version": "1.0.0",
        },
      },
    ],
  },
]
`;

exports[`MetricsSubmitter should handles updowncounter negative values correctly 1`] = `
[
  {
    "resource": ResourceImpl {
      "_asyncAttributesPending": false,
      "_memoizedAttributes": undefined,
      "_rawAttributes": [
        [
          "service.name",
          "test-service",
        ],
        [
          "service.version",
          "1.0.0",
        ],
      ],
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": -5,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.updown",
              "type": "UP_DOWN_COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": false,
          },
        ],
        "scope": {
          "name": "test-service",
          "schemaUrl": undefined,
          "version": "1.0.0",
        },
      },
    ],
  },
]
`;

exports[`MetricsSubmitter should reuses existing instruments for same metric name 1`] = `
[
  {
    "resource": ResourceImpl {
      "_asyncAttributesPending": false,
      "_memoizedAttributes": undefined,
      "_rawAttributes": [
        [
          "service.name",
          "test-service",
        ],
        [
          "service.version",
          "1.0.0",
        ],
      ],
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "test.name": "example",
                },
                "value": 3,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "Test metric",
              "name": "test-namespace.v1.test.counter",
              "type": "COUNTER",
              "unit": "{test}",
              "valueType": 1,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "test-service",
          "schemaUrl": undefined,
          "version": "1.0.0",
        },
      },
    ],
  },
]
`;
